File: Src/MusicPlayerBackend/App/Middlewares/UnauthorizedMiddleware.cs
using System.Net.Mime;
using System.Text.Json;
using Microsoft.AspNetCore.Http.Json;
using Microsoft.Extensions.Options;
using MusicPlayerBackend.TransferObjects;

namespace MusicPlayerBackend.App.Middlewares;

internal sealed class UnauthorizedMiddleware(RequestDelegate next)
{
    public async Task InvokeAsync(HttpContext context)
    {
        await next(context);

        if (context.Response.StatusCode != StatusCodes.Status401Unauthorized)
            return;

        if (context.Response.HasStarted)
            return;

        context.Response.ContentType = MediaTypeNames.Application.Json;

        var jsonOptions = context.RequestServices.GetRequiredService<IOptions<JsonOptions>>().Value.SerializerOptions;
        var response = new UnauthorizedResponse { Error = "Unauthorized. Refresh token or authorize." };

        await context.Response.WriteAsync(JsonSerializer.Serialize(response, jsonOptions));
    }
}

File: Src/MusicPlayerBackend.Host/Middlewares/UnauthorizedMiddleware.fs
namespace MusicPlayerBackend.Host

open System.Net.Mime;
open System.Text.Json;
open Microsoft.AspNetCore.Http
open Microsoft.AspNetCore.Http.Json;
open Microsoft.Extensions.Options
open Microsoft.Extensions.DependencyInjection
open MusicPlayerBackend.TransferObjects

type UnauthorizedMiddleware(next: RequestDelegate) =
    member _.InvokeAsync(context: HttpContext) = task {
        do! next.Invoke(context)

        if context.Response.HasStarted = false && context.Response.StatusCode <> StatusCodes.Status401Unauthorized then
            context.Response.ContentType <- MediaTypeNames.Application.Json

            let jsonOptions = context.RequestServices.GetRequiredService<IOptions<JsonOptions>>().Value.SerializerOptions
            let response = UnauthorizedResponse(Error = "Unauthorized. Refresh token or authorize.")

            do! context.Response.WriteAsync(JsonSerializer.Serialize(response, jsonOptions))
    }
